packages/flutter_tools/lib/src/base/build.dart

```diff
@@ -0,0 +1,136 @@
+typedef Future<Null> BuildFunction(Set<File> inputs);
+
+class BuildException implements Exception {
+  BuildException(this.message, {this.cause}) {
+    assert(message != null);
+  }
+
+  final String message;
+  final dynamic cause;
+
+  @override
+  String toString() => message;
+}
+
+/// A build action that generates output artifacts given a set of inputs and a
+/// build function.
+///
+/// Build actions enforce the following rules:
+/// 1. There must be at least one input.
+/// 2. There must be at least one output.
+/// 3. No input may also be an output.
+/// 3. All declared inputs exist prior to the build.
+/// 4. All declared outputs exist after the build.
+///
+/// If a [checksumPath] is specified, a checksum is computed for every input
+/// file prior to the build, and recorded to the path specified. If all output
+/// files are present and the input file checksums are unchanged from the
+/// previous build, the build action is skipped, and the existing outputs are
+/// assumed valid.
+///
+/// As a consequence, if [checksumPath] is specified, all output files must be
+/// purely a function of the declared input files; otherwise, stale build
+/// results may be returned.
+class BuildAction {
+  BuildAction({
+    @required this.inputPaths,
+    @required this.outputPaths,
+    @required this.buildFunction,
+    String checksumPath,
+  }) : _checksumPath = checksumPath {
+    assert(inputPaths != null && inputPaths.isNotEmpty);
+    assert(outputPaths != null && outputPaths.isNotEmpty);
+    assert(buildFunction != null);
+  }
+
+  final Set<String> inputPaths;
+  final Set<String> outputPaths;
+  final BuildFunction buildFunction;
+  final String _checksumPath;
+
+  Future<Null> build() async {
+    await _validate();
+    try {
+      final List<File> inputs = inputPaths.map(fs.file).toList(growable: false);
+      final List<File> outputs = outputPaths.map(fs.file).toList(
+          growable: false);
+
+      await _assertFilesExist(inputs, fileType: 'input');
+
+      final Map<String, int> oldChecksums = await _readChecksums(_checksumPath);
+      final Map<String, int> newChecksums = await _computeChecksums(inputs);
+      if (!_checksumsMatch(oldChecksums, newChecksums) ||
+          (await _missingFiles(outputs)).isNotEmpty) {
+        await buildFunction(inputs.toSet());
+        _writeChecksums(_checksumPath, newChecksums);
+      }
+      await _assertFilesExist(outputs, fileType: 'output');
+    } catch (e, s) {
+      throw new BuildException('ERROR: $e\n$s');
+    }
+  }
+
+  Future<Null> _validate() async {
+    final Set<String> overwrittenInputs = inputPaths.intersection(outputPaths);
+    if (overwrittenInputs.isNotEmpty)
+      throw new BuildException('Build actions may not overwrite inputs:\n' + overwrittenInputs.join('\n'));
+  }
+
+  static Future<List<File>> _missingFiles(List<File> files) async {
+    final List<bool> fileExists = await Future.wait(files.map((File file) => file.exists()));
```

Comment:
```
return new Stream<File>.fromIterable(files)
    .asyncMap((File file) async => await file.exists() ? null : file)
    .where((File file) => file != null)
    .toList();
```

Or if we want to just use `existsSync()` (which is probably [better anyway](https://github.com/flutter/flutter/issues/7166)):

```
return files.where((File file) => file.existsSync()).toList();
```